File: ex20/Xscribe.py
import os
import whisper
import srt
import argparse
from datetime import timedelta
from pyannote.audio import Pipeline

# Ensure the weights directory exists
os.makedirs("weights", exist_ok=True)

def transcribe_audio_with_diarization(input_audio_path, output_srt_path, language_code="bn"):
    # Load the Whisper model (it will download weights if not already present)
    print("Loading Whisper model...")
    model = whisper.load_model("large", download_root="weights")
    model = model.to("cpu")  # Move the model to CPU

    # Load the pyannote.audio speaker diarization pipeline
    print("Loading speaker diarization model...")
    diarization_pipeline = Pipeline.from_pretrained("pyannote/speaker-diarization")

    # Perform speaker diarization
    print("Running speaker diarization...")
    diarization = diarization_pipeline(input_audio_path)

    # Transcribe the audio file
    print("Transcribing audio...")
    transcription = model.transcribe(input_audio_path, language=language_code)

    # Align diarization results with transcription segments
    subtitles = []
    speaker_labels = {}
    speaker_counter = 1

    for segment in transcription["segments"]:
        start_time = timedelta(seconds=segment["start"])
        end_time = timedelta(seconds=segment["end"])
        text = segment["text"]

        # Find the speaker for this segment
        speaker = None
        for turn, _, speaker_id in diarization.itertracks(yield_label=True):
            if turn.start <= segment["start"] <= turn.end:
                speaker = speaker_id
                break

        # Assign a speaker label (e.g., Speaker 01, Speaker 02)
        if speaker not in speaker_labels:
            speaker_labels[speaker] = f"Speaker {speaker_counter:02d}"
            speaker_counter += 1

        # Add the speaker label to the subtitle text
        subtitle_text = f"{speaker_labels[speaker]}: {text}"

        # Create the subtitle
        subtitle = srt.Subtitle(
            index=len(subtitles) + 1,
            start=start_time,
            end=end_time,
            content=subtitle_text
        )
        subtitles.append(subtitle)

    # Write the SRT file
    with open(output_srt_path, "w", encoding="utf-8") as f:
        f.write(srt.compose(subtitles))

    print(f"SRT file generated successfully at {output_srt_path}")

if __name__ == "__main__":
    # Set up argument parsing
    parser = argparse.ArgumentParser(description="Transcribe audio with speaker diarization and generate an SRT file.")
    parser.add_argument("-i", "--input", required=True, help="Path to the input audio file.")
    parser.add_argument("-o", "--output", required=True, help="Path for the output SRT file.")
    parser.add_argument("-l", "--language", default="bn", help="Language code for transcription (e.g., 'bn' for Bengali).")
    args = parser.parse_args()

    # Transcribe and generate SRT with speaker diarization
    transcribe_audio_with_diarization(args.input, args.output, args.language)

File: ex20/Xscribe.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>
#include <chrono>
#include <iomanip>
#include <cstdlib>
#include <cstring>

// Placeholder for GGUF-based transcription and diarization
// These functions would be implemented using GGUF files and appropriate C++ libraries
std::vector<std::tuple<double, double, std::string>> transcribe_audio(const std::string& audio_path, const std::string& language_code) {
    // Placeholder for transcription logic using GGUF files
    // This would return a vector of (start_time, end_time, text) tuples
    return {{0.0, 1.0, "Hello, world!"}, {1.0, 2.0, "How are you?"}};
}

std::vector<std::tuple<double, double, std::string>> perform_diarization(const std::string& audio_path) {
    // Placeholder for diarization logic using GGUF files
    // This would return a vector of (start_time, end_time, speaker_id) tuples
    return {{0.0, 1.0, "Speaker 1"}, {1.0, 2.0, "Speaker 2"}};
}

std::string format_time(double seconds) {
    int hours = static_cast<int>(seconds) / 3600;
    int minutes = (static_cast<int>(seconds) % 3600) / 60;
    int secs = static_cast<int>(seconds) % 60;
    int millis = static_cast<int>((seconds - static_cast<int>(seconds)) * 1000);
    std::ostringstream oss;
    oss << std::setfill('0') << std::setw(2) << hours << ":"
        << std::setw(2) << minutes << ":" << std::setw(2) << secs << ","
        << std::setw(3) << millis;
    return oss.str();
}

void generate_srt(const std::string& output_path, const std::vector<std::tuple<double, double, std::string>>& transcription, const std::vector<std::tuple<double, double, std::string>>& diarization) {
    std::ofstream srt_file(output_path);
    if (!srt_file.is_open()) {
        std::cerr << "Failed to open SRT file for writing." << std::endl;
        return;
    }

    std::map<std::string, std::string> speaker_labels;
    int speaker_counter = 1;

    for (size_t i = 0; i < transcription.size(); ++i) {
        double start_time = std::get<0>(transcription[i]);
        double end_time = std::get<1>(transcription[i]);
        std::string text = std::get<2>(transcription[i]);

        std::string speaker;
        for (const auto& diar : diarization) {
            if (std::get<0>(diar) <= start_time && start_time <= std::get<1>(diar)) {
                speaker = std::get<2>(diar);
                break;
            }
        }

        if (speaker_labels.find(speaker) == speaker_labels.end()) {
            speaker_labels[speaker] = "Speaker " + std::to_string(speaker_counter++);
        }

        std::string subtitle_text = speaker_labels[speaker] + ": " + text;

        srt_file << i + 1 << "\n";
        srt_file << format_time(start_time) << " --> " << format_time(end_time) << "\n";
        srt_file << subtitle_text << "\n\n";
    }

    srt_file.close();
    std::cout << "SRT file generated successfully at " << output_path << std::endl;
}

void print_usage(const char* program_name) {
    std::cerr << "Usage: " << program_name << " -i <input_audio_file> -o <output_srt_file> [-l <language_code>]" << std::endl;
    std::cerr << "Options:" << std::endl;
    std::cerr << "  -i, --input    Path to the input audio file (required)" << std::endl;
    std::cerr << "  -o, --output   Path for the output SRT file (required)" << std::endl;
    std::cerr << "  -l, --language Language code for transcription (default: 'bn' for Bengali)" << std::endl;
    std::cerr << "  -h, --help     Display this help message" << std::endl;
}

int main(int argc, char* argv[]) {
    std::string input_audio_path;
    std::string output_srt_path;
    std::string language_code = "bn"; // Default language code is Bengali

    if (argc < 2) {
        print_usage(argv[0]);
        return 1;
    }

    // Manual argument parsing
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        
        if (arg == "-h" || arg == "--help") {
            print_usage(argv[0]);
            return 0;
        } else if (arg == "-i" || arg == "--input") {
            if (i + 1 < argc) {
                input_audio_path = argv[++i];
            } else {
                std::cerr << "Error: -i/--input requires an argument." << std::endl;
                print_usage(argv[0]);
                return 1;
            }
        } else if (arg == "-o" || arg == "--output") {
            if (i + 1 < argc) {
                output_srt_path = argv[++i];
            } else {
                std::cerr << "Error: -o/--output requires an argument." << std::endl;
                print_usage(argv[0]);
                return 1;
            }
        } else if (arg == "-l" || arg == "--language") {
            if (i + 1 < argc) {
                language_code = argv[++i];
            } else {
                std::cerr << "Error: -l/--language requires an argument." << std::endl;
                print_usage(argv[0]);
                return 1;
            }
        } else {
            std::cerr << "Unknown option: " << arg << std::endl;
            print_usage(argv[0]);
            return 1;
        }
    }

    // Check for required arguments
    if (input_audio_path.empty()) {
        std::cerr << "Error: Input audio path is required." << std::endl;
        print_usage(argv[0]);
        return 1;
    }

    if (output_srt_path.empty()) {
        std::cerr << "Error: Output SRT path is required." << std::endl;
        print_usage(argv[0]);
        return 1;
    }

    // Perform transcription and diarization
    auto transcription = transcribe_audio(input_audio_path, language_code);
    auto diarization = perform_diarization(input_audio_path);

    // Generate SRT file
    generate_srt(output_srt_path, transcription, diarization);

    return 0;
}